dens-validation/src/MyLib.hs

```diff
@@ -1,41 +1,276 @@
+            :--> PUnit
+        )
+mkSetValidator = phoistAcyclic $ plam $ \protocolAddr protocolCS setInsert cxt -> P.do
+    -- We only need to check whether the setElem token is minted, all the real validation logic is in that MP
+    txInfo <- plet $ pfield @"txInfo" # cxt
+
+    fields <- pletFields @'["inputs", "referenceInputs", "outputs", "mint"] txInfo
+
+    Protocol elemIdMP setElemMP setValidator _ <-
+        pmatch $
+            extractDatum @Protocol
+                # protocolAddr
+                # protocolCS
+                #$ pmap
+                # resolved
+                # fields.referenceInputs
+
+    setElemIdCS <- plet $ scriptHashToCS # pfromData setElemMP
+
+    pguardM "ElementID token minted" $
+        (pvalueOf # fields.mint # setElemIdCS # pcon (PTokenName $ pconstant ""))
+            #== 1
+
+    pcon PUnit
+
+mkSetElemMintingPolicy ::
+    Term
+        s
+        ( PAddress -- Address of the (unspendable) protocol validator
+            :--> PCurrencySymbol -- Protocol NFT (assume empty token name for now)
+            :--> SetInsert
+            :--> PScriptContext
+            :--> PUnit
+        )
+mkSetElemMintingPolicy = phoistAcyclic $ plam $ \protocolAddr protocolSymb setInsert cxt -> P.do
+    SetInsert'Insert keyToInsert <- pmatch setInsert
+
+    txInfo <- plet $ pfield @"txInfo" # cxt
+
+    fields <- pletFields @'["inputs", "referenceInputs", "outputs", "mint"] txInfo
+
+    -- 0) REFERENCE INPUTS: Get the protocol datum (reference input, paid to protocol validator)
+
+    Protocol elemIdMP setElemMP setValidator _ <-
+        pmatch $
+            extractDatum @Protocol
+                # protocolAddr
+                # protocolSymb
+                #$ pmap
+                # resolved
+                # fields.referenceInputs
+
+    -- 1) INPUTS: Check for k < densKey < nxt input that pays to the set validator & holds a
+
+    -- not strictly necessary but can't hurt
+    pguardM "Only one input" (plength # pfromData fields.inputs #== 1)
+    setValidatorAddress <- plet $ pscriptHashAddress # setValidator
+    setElemCS <- plet $ scriptHashToCS # pfromData setElemMP -- tx in inputs should have SetElem NFT (right?)
+    setDatum <- plet $ extractDatum @SetDatum # setValidatorAddress # setElemCS #$ pmap # resolved # fields.inputs -- ignoring OwnerApproval for now
+    SetDatum l r _ <- pmatch setDatum
+    pguardM "Validate set insert" $ validateSetInsert # setDatum # pfromData keyToInsert
+
+    -- 2) OUTPUTS: Check for a. SD(k, densKey) b. SD(densKey,nxt)
+    outputs <- plet $ pfromData fields.outputs
+    pguardM "Only two outputs" (plength # outputs #== 2)
+    l' <- plet $ pfromData l
+    r' <- plet $ pfromData r
+    k <- plet $ pfromData keyToInsert
+    checkOutput <- plet $ plam $ \p -> extractDatumSuchThat @SetDatum # p # setValidatorAddress # setElemCS # outputs
+    SetDatum{} <- pmatch $ checkOutput # (hasLR # l' # k)
+    SetDatum{} <- pmatch $ checkOutput # (hasLR # k # r')
```

Comment:
I thought we couldn't rely upon the outputs having a specific order/specific indices? Again, it's been a while since I've worked on Cardano, could be wrong (or that could have changed)

Reply:
Er, I might actually be wrong about this, although I'm pretty certain that we can build the transactions with outputs in a particular order designed specifically for the onchain code.

We can't rely on the transaction inputs since they're a set (iirc. they're sorted lexicographically based on their TxId + a little more), but the transaction outputs are a _list_ so the order suddenly matters for the ledger and I don't think there are any ledger level rules enforcing a particular order for them.